### PR TITLE
Refactor pi gate

### PR DESCRIPTION
**Pull request by ChihChengLiang**
- Fix part of #219, take and return [(Cell, F); 25] in assign region
- Achieve Pi using only copy constraints.
- Use layouter to assign a distinct region for it.

### CODE REVIEW

File: keccak256/src/gates/pi.rs [hunk 1]
```diff
@@ -12,66 +9,50 @@ use std::marker::PhantomData;
 
 #[derive(Clone, Debug)]
 pub struct PiConfig<F> {
-    q_enable: Selector,
     state: [Column<Advice>; 25],
     _marker: PhantomData<F>,
 }
 
 impl<F: FieldExt> PiConfig<F> {
-    pub const OFFSET: usize = 2;
```
Comment (CPerezz) [inline keccak256/src/gates/pi.rs:21]: Please. Don't remove the OFFSET as it's required later for assembliyng the entire Keccak circuit.

File: keccak256/src/gates/pi.rs [hunk 2]
```diff
@@ -12,66 +9,50 @@ use std::marker::PhantomData;
...
-                    let lane = meta.query_advice(
-                        state[5 * ((x + 3 * y) % 5) + x],
-                        Rotation::cur(),
-                    );
-                    let new_lane =
-                        meta.query_advice(state[5 * x + y], Rotation::next());
-
-                    q_enable.clone() * (new_lane - lane)
-                })
-                .collect::<Vec<Expression<F>>>()
```
Comment (CPerezz) [inline keccak256/src/gates/pi.rs:41]: Why are you removing the gate from Pi??
It does not apply any constraints now.

File: keccak256/src/gates/pi.rs (hunk 1 above)
Comment (ChihChengLiang) [reply keccak256/src/gates/pi.rs:21]: Sorry, I'll return them back.

Comment (ChihChengLiang) [reply keccak256/src/gates/pi.rs:21]: Oh, I realize it's part of the layouter refactor. 

Currently the circuit looks like this
```
  state = {
                // Apply pi outside circuit
                let out_state = KeccakFArith::pi(&state_to_biguint(state));
                let out_state = state_bigint_to_pallas(out_state);
                // assignment
                self.pi_config
                    .assign_state(region, offset, state, out_state)?
            };

            offset += PiConfig::OFFSET;
```
With the new interface, we can do
```
  state = self.pi_config.assign_state(mut layouter, state, out_state)?;
```
No need to track the region and the offset. The layouter find a new region for it.
What do you think?

Comment (CPerezz) [reply keccak256/src/gates/pi.rs:21]: hmmm. 

So the proposal is to have each gate/config application within a different region then?

File: keccak256/src/gates/pi.rs (hunk 2 above)
Comment (CPerezz) [reply keccak256/src/gates/pi.rs:41]: Hmmm. But copy constraints can only enforce that the provided input is indeed the one that has to be used.

Not sure how you still enforce with constraints:
```rust
 let lane = meta.query_advice(
                        state[5 * ((x + 3 * y) % 5) + x],
                        Rotation::cur(),
                    );
                    let new_lane =
                        meta.query_advice(state[5 * x + y], Rotation::next());

                    q_enable.clone() * (new_lane - lane)
```

File: keccak256/src/gates/pi.rs (hunk 1 above)
Comment (ChihChengLiang) [reply keccak256/src/gates/pi.rs:21]: Yes, it makes sense to let layouter decide where to put the state row. Counting offsets is like doing the layouter's job.

File: keccak256/src/gates/pi.rs (hunk 2 above)
Comment (ChihChengLiang) [reply keccak256/src/gates/pi.rs:41]: IIU the halo2 code correctly, copy constraints are applicable as long as we know what cells to copy. We don't have to create custom gates for them.
I did realize I forgot to add enable equality in this case. Will add now.

Comment (CPerezz) [reply keccak256/src/gates/pi.rs:41]: Ohh I see! That sounds correct! Thanks!

**Review by CPerezz — APPROVED** [review]
LGTM! 
Depending on the `Layouter` instead of `Region + offset` seems like a really nice idea!

**Comment by ChihChengLiang** [comment]
Close in favor of #234

**Pull request closed without merging**